online-inference/gpt-j-deepspeed/service/service.py

```diff
@@ -0,0 +1,292 @@
+                    )
+                    request_params[pk] = pv
+
+        if "benchmark" in request:
+            request_params["BENCHMARK"] = True
+            request_params["MIN_LENGTH"] = request_params["MAX_LENGTH"]
+        else:
+            request_params["BENCHMARK"] = False
+
+        if "instances" in request:
+            request_params["INSTANCES"] = request["instances"]
+
+        return request_params
+
+    def _predict(self, request_params):
+        return {
+            "predictions": self.generator(
+                request_params["INSTANCES"][0],
+                do_sample=False,
+                min_length=request_params["MIN_LENGTH"],
+                max_length=request_params["MAX_LENGTH"],
+                temperature=request_params["TEMPERATURE"],
+                top_k=request_params["TOP_K"],
+                top_p=request_params["TOP_P"],
+                repetition_penalty=request_params["REPETITION_PENALTY"],
+                pad_token_id=self.tokenizer.eos_token_id,
+            )
+        }
+
+    def predict(self, request: Dict) -> Dict:
+        request_params = self.configure_request(request)
+        if request_params["BENCHMARK"] is True:
+            return self.benchmark(request_params)
+        else:
+            return self._predict(request_params)
+
+    def load_dataset(self):
+        data = open("the-time-machine.txt", "r").read()
+        tokens = self.tokenizer.encode(data)
+        sequence_start = random.randrange(len(tokens) - 2048)
+        sequence_end = sequence_start + 2048
+        return tokens[sequence_start:sequence_end]
+
+    def benchmark(self, request_params):
+        if self.dataset is None:
+            self.dataset = self.load_dataset()
+
+        assert (
+            request_params["BENCHMARK_SEQUENCE_LENGTH"] < request_params["MAX_LENGTH"]
+        )
+
+        sequence_start = random.randrange(
+            len(self.dataset) - request_params["BENCHMARK_SEQUENCE_LENGTH"]
+        )
+        sequence_end = sequence_start + request_params["BENCHMARK_SEQUENCE_LENGTH"]
+        random_sequence_encoded = self.dataset[sequence_start:sequence_end]
+        random_sequence = self.tokenizer.decode(random_sequence_encoded)
+
+        request_params["INSTANCES"] = [random_sequence]
+
+        start = time.time()
+        predicitions = self._predict(request_params)
+        end = time.time()
+        generation_time = end - start
+
+        logger.info(
+            f'Tokens In: {request_params["BENCHMARK_SEQUENCE_LENGTH"]}, New Tokens: {request_params["MAX_LENGTH"]- request_params["BENCHMARK_SEQUENCE_LENGTH"]}, Generation Time: {generation_time}'
+        )
+
+        return {
+            "benchmark_results": {
+                "input_sequence_length": request_params["BENCHMARK_SEQUENCE_LENGTH"],
+                "generated_tokens": request_params["MAX_LENGTH"]
+                - request_params["BENCHMARK_SEQUENCE_LENGTH"],
+                "time": generation_time,
+            }
```

Comment:
I think iterating over the sequences will change this. You can write to a JSON (append) and read at the end to get the quantiles for the different number of experiments you run.